lib/platform-stats.js

```diff
@@ -165,37 +271,168 @@ export const mapParticipantsToIds = async (pgClient, participantsSet) => {
     ON CONFLICT(participant_address) DO UPDATE
-      -- this no-op update is needed to populate "RETURNING id"
+      -- this no-op update is needed to populate "RETURNING id, participant_address"
       SET participant_address = EXCLUDED.participant_address
-    RETURNING id
+    RETURNING id, participant_address
   `, [
     newAddresses
   ])
 
   assert.strictEqual(created.length, newAddresses.length)
-  return ids.concat(created.map(r => r.id))
+  for (const { id, participant_address } of created) {
+    participantMap.set(participant_address, id)
+  }
+
+  return participantMap;
+}
+
+export const periodicDatabaseRefresh = async (createPgClient) => {
+  const pgClient = await createPgClient();
+  try {
+    await updateTopMeasurementParticipants(pgClient);
+    await aggregateAndCleanupRecentData(pgClient);
+    await updateMonthlyActiveStationCount(pgClient);
+  } finally {
+    await pgClient.end();
+  }
 }
 
 /**
- * @param {import('./typings.js').CreatePgClient} createPgClient
+ * @param {import('pg').Client} pgClient
  */
-export const updateTopMeasurementParticipants = async (createPgClient) => {
+export const updateTopMeasurementParticipants = async (pgClient) => {
   try {
     await pRetry(async () => {
-      const pgClient = await createPgClient()
-      try {
-        await pgClient.query('REFRESH MATERIALIZED VIEW top_measurement_participants_yesterday_mv')
-      } finally {
-        await pgClient.end()
-      }
+      await pgClient.query('REFRESH MATERIALIZED VIEW top_measurement_participants_yesterday_mv');
     }, {
       retries: 3,
       onFailedAttempt: (error) => {
-        console.warn(`Attempt to refresh materialized view failed: ${error.message}. Retrying...`)
+        console.warn(`Attempt to refresh materialized view failed: ${error.message}. Retrying...`);
       }
-    })
+    });
   } catch (err) {
-    console.error('Error refreshing top measurement participants', err)
-    Sentry.captureException(err)
+    console.error('Error refreshing top measurement participants', err);
+    Sentry.captureException(err);
   }
 }
+
+/**
+ * @param {import('pg').Client} pgClient
+ */
+const aggregateAndCleanupRecentData = async (pgClient) => {
+  try {
+    const twoDaysAgo = new Date();
+    twoDaysAgo.setDate(twoDaysAgo.getDate() - 2);
+
+    await pgClient.query('BEGIN');
+
+    // The ON CONFLICT theoretically should not happen, because it's two days old data,
+    // which should not be updated anymore. But it's a good practice to have it.
+    await pgClient.query(`
+      WITH rsdn_summary AS (
+        SELECT
+          day,
+          SUM(accepted_measurement_count) AS accepted_measurement_count,
+          SUM(total_measurement_count) AS total_measurement_count,
+          COUNT(DISTINCT station_id) AS distinct_active_station_count,
+          COUNT(DISTINCT participant_id) AS distinct_participant_address_count
+        FROM
+          recent_station_details
+        WHERE
+          day <= $1::DATE
```

Comment:
As an alternative to calculating "two days ago" in JavaScript, I think we should be able to use something like `now() - INTERVAL '2 days'` or `now()::DATE - 'INTERVAL '2 days'`.

We need to be careful about time zones - 1 am on Tuesday in Prague is 7 pm on Monday in New York. I expect that all our services are running in UTC time, so we should be fine.

Reply:
Addressed in https://github.com/filecoin-station/spark-evaluate/pull/311/commits/bc1515657bc9a0209a1890df58c19734ec31d6ff

I also did the same for the lastMonth calculation

I agree that timezones should theoretically not mess us up here; I think we do similar things today with dashboard code